parametric/test_headers_tracecontext.py

```diff
@@ -0,0 +1,907 @@
+    assert "foo=1" in str(tracestate1)
+
+    assert traceparent2.trace_id.hex() == "12345678901234567890123456789012"
+    assert "foo=1" in str(tracestate2) or "foo=2" in str(tracestate2)
+
+    assert traceparent3.trace_id.hex() == "12345678901234567890123456789012"
+    assert "foo=1" in str(tracestate3)
+
+    assert traceparent4.trace_id.hex() == "12345678901234567890123456789012"
+    assert "foo=1" in str(tracestate4) or "foo=2" in str(tracestate4)
+
+
+@temporary_enable_optin_tracecontext()
+@pytest.mark.skip_library("dotnet", "Tracestate not implemented")
+@pytest.mark.skip_library("golang", "not implemented")
+@pytest.mark.skip_library("nodejs", "not implemented")
+@pytest.mark.skip_library("python", "not implemented")
+def test_tracestate_all_allowed_characters(test_agent, test_library):
+    """
+    harness sends a request with a valid tracestate header with all legal characters
+    expects the tracestate to be inherited
+    """
+    key_without_vendor = "".join(
+        ["".join(map(chr, range(0x61, 0x7A + 1))), "0123456789", "_", "-", "*", "/",]  # lcalpha  # DIGIT
+    )
+    key_with_vendor = key_without_vendor + "@a-z0-9_-*/"
+    value = "".join(
+        [
+            "".join(map(chr, range(0x20, 0x2B + 1))),
+            "".join(map(chr, range(0x2D, 0x3C + 1))),
+            "".join(map(chr, range(0x3E, 0x7E + 1))),
+        ]
+    )
+
+    with test_library:
+        traceparent1, tracestate1 = make_single_request_and_get_tracecontext(
+            test_library,
+            [
+                ["traceparent", "00-12345678901234567890123456789012-1234567890123456-00"],
+                ["tracestate", key_without_vendor + "=" + value],
+            ],
+        )
+
+        traceparent2, tracestate2 = make_single_request_and_get_tracecontext(
+            test_library,
+            [
+                ["traceparent", "00-12345678901234567890123456789012-1234567890123456-00"],
+                ["tracestate", key_with_vendor + "=" + value],
+            ],
+        )
+
+    assert key_without_vendor in tracestate1
+    assert tracestate1[key_without_vendor] == value
+
+    assert key_with_vendor in tracestate2
+    assert tracestate2[key_with_vendor] == value
+
+
+@temporary_enable_optin_tracecontext()
+@pytest.mark.skip_library("dotnet", "Tracestate not implemented")
+@pytest.mark.skip_library("golang", "not implemented")
+@pytest.mark.skip_library("nodejs", "not implemented")
+@pytest.mark.skip_library("python", "not implemented")
+def test_tracestate_ows_handling(test_agent, test_library):
+    """
+    harness sends a request with a valid tracestate header with OWS
+    expects the tracestate to be inherited
+    """
+    with test_library:
+        traceparent1, tracestate1 = make_single_request_and_get_tracecontext(
+            test_library,
+            [
+                ["traceparent", "00-12345678901234567890123456789012-1234567890123456-00"],
+                ["tracestate", "foo=1 \t , \t bar=2, \t baz=3"],
```

Comment:
According to the w3c spec tabs are invalid value: https://www.w3.org/TR/trace-context/#value. I think we need use another character in this test

Reply:
This test case came from the w3c/tracecontext so it should be valid 😆

And looking at [this grammar](https://www.w3.org/TR/trace-context/#list), it does appear to be valid because:
1. OWS (optional whitespace) is allowed between the previous comma and the next `key=value`
2. OWS (optional whitespace) is allowed after a `key=value` and before the next comma

Reply:
Also [in this section](https://www.w3.org/TR/trace-context/#tracestate-header-field-values) it asserts the following:

> Spaces and horizontal tabs surrounding list-members are ignored.

I think what would be invalid was if there was a \t in the middle of a value, like `foo=123\t456`

Reply:
Resolving, but let me know if you have further questions